File: python/api/embeddings.py
"""
Embeddings API endpoints
"""

import logging
from typing import Optional, List
from fastapi import APIRouter, HTTPException
from pydantic import BaseModel

from services.embeddings import embeddings_service
from utils.logger import get_logger

logger = get_logger(__name__)

router = APIRouter()


class EmbeddingRequest(BaseModel):
    """Request model for single embedding generation."""
    text: str


class EmbeddingsRequest(BaseModel):
    """Request model for batch embedding generation."""
    texts: List[str]


class EmbeddingResponse(BaseModel):
    """Response model for single embedding."""
    embedding: List[float]
    dimension: int


class EmbeddingsResponse(BaseModel):
    """Response model for batch embeddings."""
    embeddings: List[List[float]]
    dimension: int
    count: int


class SimilarityRequest(BaseModel):
    """Request model for similarity computation."""
    embedding1: List[float]
    embedding2: List[float]


class SimilarityResponse(BaseModel):
    """Response model for similarity computation."""
    similarity: float


class SimilaritySearchRequest(BaseModel):
    """Request model for similarity search."""
    query_embedding: List[float]
    candidate_embeddings: List[List[float]]
    top_k: Optional[int] = 5


class SimilaritySearchResponse(BaseModel):
    """Response model for similarity search."""
    results: List[dict]


@router.post("/generate", response_model=EmbeddingResponse)
async def generate_embedding(request: EmbeddingRequest):
    """
    Generate embedding for a single text.
    
    Args:
        request: Embedding request with text
    
    Returns:
        Generated embedding
    """
    try:
        if not await embeddings_service.is_ready():
            raise HTTPException(status_code=503, detail="Embeddings service not ready")
        
        if not request.text.strip():
            raise HTTPException(status_code=400, detail="Text is required")
        
        # Generate embedding
        embedding = await embeddings_service.generate_embedding(request.text)
        
        if embedding is None:
            raise HTTPException(status_code=500, detail="Failed to generate embedding")
        
        return EmbeddingResponse(
            embedding=embedding,
            dimension=len(embedding)
        )
    
    except Exception as e:
        logger.error(f"Embedding generation failed: {e}")
        raise HTTPException(status_code=500, detail=f"Embedding generation failed: {str(e)}")


@router.post("/generate-batch", response_model=EmbeddingsResponse)
async def generate_embeddings(request: EmbeddingsRequest):
    """
    Generate embeddings for multiple texts.
    
    Args:
        request: Batch embedding request with texts
    
    Returns:
        Generated embeddings
    """
    try:
        if not await embeddings_service.is_ready():
            raise HTTPException(status_code=503, detail="Embeddings service not ready")
        
        if not request.texts or not any(text.strip() for text in request.texts):
            raise HTTPException(status_code=400, detail="At least one non-empty text is required")
        
        # Generate embeddings
        embeddings = await embeddings_service.generate_embeddings(request.texts)
        
        if embeddings is None:
            raise HTTPException(status_code=500, detail="Failed to generate embeddings")
        
        dimension = len(embeddings[0]) if embeddings else 0
        
        return EmbeddingsResponse(
            embeddings=embeddings,
            dimension=dimension,
            count=len(embeddings)
        )
    
    except Exception as e:
        logger.error(f"Batch embedding generation failed: {e}")
        raise HTTPException(status_code=500, detail=f"Batch embedding generation failed: {str(e)}")


@router.post("/similarity", response_model=SimilarityResponse)
async def compute_similarity(request: SimilarityRequest):
    """
    Compute cosine similarity between two embeddings.
    
    Args:
        request: Similarity request with two embeddings
    
    Returns:
        Similarity score
    """
    try:
        if not await embeddings_service.is_ready():
            raise HTTPException(status_code=503, detail="Embeddings service not ready")
        
        if not request.embedding1 or not request.embedding2:
            raise HTTPException(status_code=400, detail="Both embeddings are required")
        
        if len(request.embedding1) != len(request.embedding2):
            raise HTTPException(status_code=400, detail="Embeddings must have the same dimension")
        
        # Compute similarity
        similarity = await embeddings_service.compute_similarity(
            request.embedding1,
            request.embedding2
        )
        
        return SimilarityResponse(similarity=similarity)
    
    except Exception as e:
        logger.error(f"Similarity computation failed: {e}")
        raise HTTPException(status_code=500, detail=f"Similarity computation failed: {str(e)}")


@router.post("/search", response_model=SimilaritySearchResponse)
async def similarity_search(request: SimilaritySearchRequest):
    """
    Find most similar embeddings to a query embedding.
    
    Args:
        request: Similarity search request
    
    Returns:
        Top similar embeddings with their indices and scores
    """
    try:
        if not await embeddings_service.is_ready():
            raise HTTPException(status_code=503, detail="Embeddings service not ready")
        
        if not request.query_embedding or not request.candidate_embeddings:
            raise HTTPException(status_code=400, detail="Query and candidate embeddings are required")
        
        # Check dimensions
        query_dim = len(request.query_embedding)
        for i, candidate in enumerate(request.candidate_embeddings):
            if len(candidate) != query_dim:
                raise HTTPException(
                    status_code=400,
                    detail=f"Candidate embedding {i} has different dimension than query"
                )
        
        # Perform similarity search
        results = await embeddings_service.find_most_similar(
            request.query_embedding,
            request.candidate_embeddings,
            request.top_k
        )
        
        return SimilaritySearchResponse(results=results)
    
    except Exception as e:
        logger.error(f"Similarity search failed: {e}")
        raise HTTPException(status_code=500, detail=f"Similarity search failed: {str(e)}")


@router.post("/test")
async def test_embeddings():
    """Test embeddings service with a simple text."""
    try:
        if not await embeddings_service.is_ready():
            raise HTTPException(status_code=503, detail="Embeddings service not ready")
        
        result = await embeddings_service.test_embedding()
        return result
    except Exception as e:
        logger.error(f"Embeddings test failed: {e}")
        raise HTTPException(status_code=500, detail=f"Embeddings test failed: {str(e)}")


@router.get("/ready")
async def check_ready():
    """Check if Embeddings service is ready."""
    ready = await embeddings_service.is_ready()
    return {"ready": ready}


@router.get("/info")
async def get_info():
    """Get Embeddings service information."""
    info = await embeddings_service.get_model_info()
    return info

File: backend/internal/api/embeddings.go
package api

import (
	"encoding/json"
	"net/http"

	"github.com/gorilla/mux"
)

// EmbeddingRequest represents a single embedding request
type EmbeddingRequest struct {
	Text string `json:"text"`
}

// EmbeddingResponse represents a single embedding response
type EmbeddingResponse struct {
	Embedding []float32 `json:"embedding"`
	Text      string    `json:"text"`
}

// BatchEmbeddingRequest represents a batch embedding request
type BatchEmbeddingRequest struct {
	Texts []string `json:"texts"`
}

// BatchEmbeddingResponse represents a batch embedding response
type BatchEmbeddingResponse struct {
	Embeddings [][]float32 `json:"embeddings"`
}

// SimilarityRequest represents a similarity computation request
type SimilarityRequest struct {
	Embedding1 []float32 `json:"embedding1"`
	Embedding2 []float32 `json:"embedding2"`
}

// SimilarityResponse represents a similarity computation response
type SimilarityResponse struct {
	Similarity float32 `json:"similarity"`
}

// SearchRequest represents a similarity search request
type SearchRequest struct {
	QueryEmbedding      []float32   `json:"query_embedding"`
	CandidateEmbeddings [][]float32 `json:"candidate_embeddings"`
	TopK                int         `json:"top_k,omitempty"`
}

// SearchResponse represents a similarity search response
type SearchResponse struct {
	Indices      []int     `json:"indices"`
	Similarities []float32 `json:"similarities"`
}

// GenerateEmbedding handles single embedding generation
func (h *Handler) GenerateEmbedding(w http.ResponseWriter, r *http.Request) {
	if !h.config.Features.Embeddings {
		h.writeError(w, http.StatusServiceUnavailable, "Embeddings service is disabled")
		return
	}

	embeddingService := h.modelManager.GetEmbeddingService()
	if embeddingService == nil || !embeddingService.IsReady() {
		h.writeError(w, http.StatusServiceUnavailable, "Embeddings service is not ready")
		return
	}

	var req EmbeddingRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		h.writeError(w, http.StatusBadRequest, "Invalid request body")
		return
	}

	if req.Text == "" {
		h.writeError(w, http.StatusBadRequest, "Text is required")
		return
	}

	embedding, err := embeddingService.GenerateEmbedding(r.Context(), req.Text)
	if err != nil {
		h.writeError(w, http.StatusInternalServerError, "Embedding generation failed: "+err.Error())
		return
	}

	h.writeSuccess(w, EmbeddingResponse{
		Embedding: embedding,
		Text:      req.Text,
	})
}

// GenerateEmbeddings handles batch embedding generation
func (h *Handler) GenerateEmbeddings(w http.ResponseWriter, r *http.Request) {
	if !h.config.Features.Embeddings {
		h.writeError(w, http.StatusServiceUnavailable, "Embeddings service is disabled")
		return
	}

	embeddingService := h.modelManager.GetEmbeddingService()
	if embeddingService == nil || !embeddingService.IsReady() {
		h.writeError(w, http.StatusServiceUnavailable, "Embeddings service is not ready")
		return
	}

	var req BatchEmbeddingRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		h.writeError(w, http.StatusBadRequest, "Invalid request body")
		return
	}

	if len(req.Texts) == 0 {
		h.writeError(w, http.StatusBadRequest, "Texts array is required")
		return
	}

	embeddings, err := embeddingService.GenerateEmbeddings(r.Context(), req.Texts)
	if err != nil {
		h.writeError(w, http.StatusInternalServerError, "Batch embedding generation failed: "+err.Error())
		return
	}

	h.writeSuccess(w, BatchEmbeddingResponse{
		Embeddings: embeddings,
	})
}

// ComputeSimilarity handles similarity computation
func (h *Handler) ComputeSimilarity(w http.ResponseWriter, r *http.Request) {
	if !h.config.Features.Embeddings {
		h.writeError(w, http.StatusServiceUnavailable, "Embeddings service is disabled")
		return
	}

	embeddingService := h.modelManager.GetEmbeddingService()
	if embeddingService == nil || !embeddingService.IsReady() {
		h.writeError(w, http.StatusServiceUnavailable, "Embeddings service is not ready")
		return
	}

	var req SimilarityRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		h.writeError(w, http.StatusBadRequest, "Invalid request body")
		return
	}

	if len(req.Embedding1) == 0 || len(req.Embedding2) == 0 {
		h.writeError(w, http.StatusBadRequest, "Both embeddings are required")
		return
	}

	similarity, err := embeddingService.ComputeSimilarity(r.Context(), req.Embedding1, req.Embedding2)
	if err != nil {
		h.writeError(w, http.StatusInternalServerError, "Similarity computation failed: "+err.Error())
		return
	}

	h.writeSuccess(w, SimilarityResponse{
		Similarity: similarity,
	})
}

// SearchSimilar handles similarity search
func (h *Handler) SearchSimilar(w http.ResponseWriter, r *http.Request) {
	if !h.config.Features.Embeddings {
		h.writeError(w, http.StatusServiceUnavailable, "Embeddings service is disabled")
		return
	}

	embeddingService := h.modelManager.GetEmbeddingService()
	if embeddingService == nil || !embeddingService.IsReady() {
		h.writeError(w, http.StatusServiceUnavailable, "Embeddings service is not ready")
		return
	}

	var req SearchRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		h.writeError(w, http.StatusBadRequest, "Invalid request body")
		return
	}

	if len(req.QueryEmbedding) == 0 {
		h.writeError(w, http.StatusBadRequest, "Query embedding is required")
		return
	}

	if len(req.CandidateEmbeddings) == 0 {
		h.writeError(w, http.StatusBadRequest, "Candidate embeddings are required")
		return
	}

	indices, similarities, err := embeddingService.SearchSimilar(
		r.Context(),
		req.QueryEmbedding,
		req.CandidateEmbeddings,
		req.TopK,
	)
	if err != nil {
		h.writeError(w, http.StatusInternalServerError, "Similarity search failed: "+err.Error())
		return
	}

	h.writeSuccess(w, SearchResponse{
		Indices:      indices,
		Similarities: similarities,
	})
}

// GetEmbeddingsInfo returns embeddings service information
func (h *Handler) GetEmbeddingsInfo(w http.ResponseWriter, r *http.Request) {
	if !h.config.Features.Embeddings {
		h.writeError(w, http.StatusServiceUnavailable, "Embeddings service is disabled")
		return
	}

	embeddingService := h.modelManager.GetEmbeddingService()
	if embeddingService == nil {
		h.writeError(w, http.StatusServiceUnavailable, "Embeddings service not available")
		return
	}

	info := embeddingService.GetInfo()
	h.writeSuccess(w, info)
}

// RegisterEmbeddingsRoutes registers embeddings-related routes
func (h *Handler) RegisterEmbeddingsRoutes(router *mux.Router) {
	embeddingsRouter := router.PathPrefix("/api/embeddings").Subrouter()

	embeddingsRouter.HandleFunc("/generate", h.GenerateEmbedding).Methods("POST")
	embeddingsRouter.HandleFunc("/batch", h.GenerateEmbeddings).Methods("POST")
	embeddingsRouter.HandleFunc("/similarity", h.ComputeSimilarity).Methods("POST")
	embeddingsRouter.HandleFunc("/search", h.SearchSimilar).Methods("POST")
	embeddingsRouter.HandleFunc("/info", h.GetEmbeddingsInfo).Methods("GET")
}
